crates/libs/bindgen/src/index.rs

```diff
@@ -0,0 +1,137 @@
+use super::*;
+use serde::Serialize;
+
+#[derive(Default, Serialize)]
+struct Index {
+    version: u8,
+    namespaces: Vec<String>,
+    items: BTreeMap<usize, Vec<IndexItem>>,
+}
+
+#[derive(Default, Serialize)]
+struct IndexItem {
+    #[serde(rename = "n")]
+    name: String,
+    #[serde(rename = "f")]
+    features: Vec<usize>,
+}
+
+impl Index {
+    fn new() -> Self {
+        Self {
+            version: 2,
+            ..Default::default()
+        }
+    }
+
+    fn get_or_create_namespace_index(&mut self, namespace: &str) -> usize {
+        match self.namespaces.iter().position(|ns| ns == namespace) {
+            Some(idx) => idx,
+            None => {
+                self.namespaces.push(namespace.to_string());
+                self.namespaces.len() - 1
+            }
+        }
+    }
+
+    fn add_item(&mut self, namespace: &str, name: &str, features: BTreeSet<String>) {
+        let namespace_idx = self.get_or_create_namespace_index(namespace);
+
+        let mut compact = BTreeSet::new();
+        for feature in features.iter().rev() {
+            if feature.is_empty() {
+                continue;
+            }
+
+            if !compact
+                .iter()
+                .any(|c: &&str| namespace_starts_with(c, feature))
+            {
+                compact.insert(feature.as_str());
+            }
+        }
+
+        let index_item = IndexItem {
+            name: name.to_string(),
+            features: compact
+                .iter()
+                .map(|dep| self.get_or_create_namespace_index(dep))
+                .collect(),
+        };
+
+        self.items
+            .entry(namespace_idx)
+            .or_default()
+            .push(index_item);
+    }
+}
+
+const EXCLUDED_NAMESPACES: &[&str] = &["Windows.Foundation", "Windows.Win32.Foundation"];
+
+#[doc(hidden)]
+pub fn write() {
+    let mut feature_index = Index::new();
+    let reader = Reader::new(expand_input(&["default"]));
```

Comment:
So this is not specific to the invocation of `bindgen` so it makes more sense to have it as a separate step. We can add a hidden `features` function that returns the string and then `tool_features` can simply call it. Or we just make the reader public.